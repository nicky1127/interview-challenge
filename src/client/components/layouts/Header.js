import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

const Header = props => {
  const { dishNum, dietariesNum } = props;

  let content;

  if (Array.isArray(dietariesNum) && dietariesNum.length > 0) {
    content = dietariesNum.map(dietary => (
      <span>
        <span>{`${dietary.num}`}</span>{" "}
        <span className="dietary">{`${dietary.dietary}`}</span>
      </span>
    ));
  }

  return (
    <div className="menu-summary" data-testid="header">
      <div className="container">
        <div className="row">
          <div className="col-6 menu-summary-left">
            <span>{`${dishNum} TOTAL DISHES`}</span>
          </div>
          <div className="col-6 menu-summary-right">{content}</div>
        </div>
      </div>
    </div>
  );
};

Header.defaultProps = { selectedItems: [] };

Header.propTypes = {
  selectedItems: PropTypes.array
};

const mapStateToProps = state => {
  const { selectedItems, items } = state;
  const dietaries = ["v", "ve", "df", "gf", "n!", "rsf"];
  const menuItems = [];

  selectedItems.forEach(el => {
    const obj = items.find(ele => ele.id === el);
    menuItems.push(obj);
  });

  const dietariesNum = [];
  dietaries.forEach(dietary => {
    let num = 0;
    menuItems.forEach(item => {
      item.dietaries.find(el => el === dietary) && num++;
    });
    num && dietariesNum.push({ dietary, num });
  });

  return {
    dishNum: selectedItems.length,
    dietariesNum
  };
};

const ConnectedHeader = connect(mapStateToProps)(Header);

export default ConnectedHeader;
