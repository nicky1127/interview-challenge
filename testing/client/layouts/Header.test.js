import { render, cleanup } from "@testing-library/react";
import Header from "client/components/layouts/Header";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

beforeEach(() => {
  jest.clearAllMocks();
  cleanup();
});

const state = {
  selectedItems: [],
  items: [{ name: "test", id: 1, dietaries: ["test"] }],
  ldgItems: false,
  ldgSuccItems: false
};
const mockStore = configureMockStore();
const store = mockStore(state);

describe("Header", () => {
  it("Snapshot matchs", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Header />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Header should render one [data-testid=Header]", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Header />
      </Provider>
    );
    const container = getByTestId("header");
    expect(container).toBeInTheDocument();
  });
});
