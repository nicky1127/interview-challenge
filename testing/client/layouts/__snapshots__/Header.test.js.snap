// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Snapshot matchs 1`] = `
<div
  className="menu-summary"
  data-testid="header"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-6 menu-summary-left"
      >
        <span>
          0 TOTAL DISHES
        </span>
      </div>
      <div
        className="col-6 menu-summary-right"
      />
    </div>
  </div>
</div>
`;
